import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/MultipleDestinationUploadService.java

```
@@ -93,26 +103,64 @@ public void start(ImmutableList<IDicomWebClient> healthcareDestinations,
+          throw new MultipleDestinationUploadServiceException(ie);
+        }
+      }
+      if (asyncUploadProcessingExceptions.isEmpty()) {
+        backupUploadService.removeBackup(sopInstanceUID);
       }
     }
 
-    if (asyncUploadProcessingExceptions.isEmpty()) {
-      backupUploadService.removeBackup(sopInstanceUID);
-    } else {
-      log.error("Exception messages of the upload async jobs:\n{}",
-          asyncUploadProcessingExceptions.stream().map(t -> t.getMessage()).collect(Collectors.joining("\n")));;
-
+    if (!asyncUploadProcessingExceptions.isEmpty()) {
+      log.error(
+          "Exception messages of the upload async jobs:\n{}",
+          asyncUploadProcessingExceptions.stream()
+              .map(t -> t.getMessage())
+              .collect(Collectors.joining("\n")));
       throw new MultipleDestinationUploadServiceException(asyncUploadProcessingExceptions.get(0));
     }
   }
+
+  @Override
+  public void cleanup(int associationId) {
+    removeLazyResolvers(associationId);
+  }
+
+  private void addLazyResolvers(
+      int associationId, String sopInstanceUID, List<CompletableFuture> futures) {
+    HashMap<String, List<CompletableFuture>> resolverMap =
+        lazyFutureResolvers.getOrDefault(
+            associationId, new HashMap<String, List<CompletableFuture>>());
+    resolverMap.put(sopInstanceUID, futures);
+    if (!lazyFutureResolvers.containsKey(associationId)) {
+      lazyFutureResolvers.put(associationId, resolverMap);
+    }
+  }
+
+  private void removeLazyResolvers(int associationId) {
+    HashMap<String, List<CompletableFuture>> resolverMap = lazyFutureResolvers.get(associationId);
+    if (resolverMap != null) {
+      log.debug("Cleaning up " + resolverMap.size() + " files for association " + associationId);
+      resolverMap.forEach(
+          (sopInstanceUID, futures) -> {
+            futures.forEach(CompletableFuture::join);
+            backupUploadService.removeBackup(sopInstanceUID);
+          });
```

This waits for the CompletableFuture to finish before removing the backup, correct?

---

Also any error reporting we want to do here if the future failed or threw an exception?

---

Yes, the original code waited for all the futures to complete. Unfortunately it silently masked exceptions, so I broke out the code to explicitly handle any exceptions, similar to the synchronous code.